backend/src/main/java/com/mcmasterbaja/analyzer/SGolayFilter.java

```diff
@@ -42,53 +77,39 @@ public void analyze() {
             + " and polynomial degree "
             + polynomialDegree);
 
-    Reader r = new CSVReader(super.inputFiles[0]);
-    Writer w = new CSVWriter(super.outputFiles[0]);
+    CSVReader reader = getReader(inputFiles[0]);
+    ICSVWriter writer = getWriter(outputFiles[0]);
 
-    w.write(sGolay(r.read(), windowSize, polynomialDegree));
-  }
-
-  public List<List<String>> sGolay(List<List<String>> data, int windowSize, int polynomialDegree) {
-    // This would literally be a one liner in python... too bad no one said we should use python
-    List<List<String>> dataPoints = new ArrayList<List<String>>();
-    List<String> dataPoint = new ArrayList<String>(2);
-
-    int independentColumn = this.getAnalysisColumnIndex(0, data.get(0));
-    int dependentColumn = this.getAnalysisColumnIndex(1, data.get(0));
-
-    // Add header
-    dataPoint.add(data.get(0).get(independentColumn));
-    dataPoint.add(data.get(0).get(dependentColumn));
-    dataPoints.add(dataPoint);
-    dataPoint = new ArrayList<String>(2);
+    String[] headers = reader.readNext();
+    int xAxisIndex = this.getColumnIndex(inputColumns[0], headers);
+    int yAxisIndex = this.getColumnIndex(inputColumns[1], headers);
+    writer.writeNext(headers);
 
-    // Put first row of data in
-    dataPoint.add(data.get(1).get(independentColumn));
-    dataPoint.add(data.get(1).get(dependentColumn));
-    dataPoints.add(dataPoint);
-    dataPoint = new ArrayList<String>(2);
-
-    int halfWindowSize = windowSize / 2;
+    String[] dataPoint;
     RealMatrix coeffMatrix = savGolCoeff(windowSize, polynomialDegree);
 
-    for (int i = 2; i < data.size(); i++) {
-      dataPoint.add(data.get(i).get(independentColumn)); // Add timestamp
+    while ((dataPoint = reader.readNext()) != null) {
+      dataBuffer.addPoint(Double.parseDouble(dataPoint[yAxisIndex]));
+      timestampBuffer.addPoint(Double.parseDouble(dataPoint[xAxisIndex]));
+
+      if (dataBuffer.size() < windowSize) {
+        continue;
+      }
 
       double smoothedValue = 0.0;
+      Double[] dataPoints = dataBuffer.getPoints();
 
-      for (int j = -halfWindowSize; j < halfWindowSize; j++) {
-        int dataIndex = Math.min(Math.max(i + j, 1), data.size() - 1);
-        double dataValue = Double.parseDouble(data.get(dataIndex).get(dependentColumn));
-        smoothedValue += coeffMatrix.getEntry(halfWindowSize + j, 0) * dataValue;
+      for (int i = 0; i < dataBuffer.size(); i++) {
+        smoothedValue += coeffMatrix.getEntry(i, 0) * dataPoints[i];
```

Comment:
Double check this

Reply:
I mean it outputted the right thing so actually nvm this is probably right

Reply:
Oh yah they equivalent